### ISSUE TITLE
OneTrainer uses like 29 GB VRAM with FP8 Qwen - Musubi uses 35 GB VRAM - What could be difference you think

### ISSUE THREAD

**Issue by FurkanGozukara** [kohya-ss/musubi-tuner#715]
I am comparing speed of OneTrainer and Musubi Tuner

Qwen Model

Both FP8 and LoRA Rank 128 - Adafactor optimizer - 1328x1328 px

OneTrainer uses 29.15 GB VRAM and fully fits into GPU

<img width="3487" height="928" alt="Image" src="https://github.com/user-attachments/assets/11a6dca0-e7c3-4635-b4b5-953ab3172b3e" />

Musubi requires 7 block swap and it uses 31.2 GB at that

For 0 block it needs like 35 GB VRAM

What could be the reason you think?

by the way speed are close. I wonder if we are using unncessary extra VRAM somewhere @kohya-ss  ?

here below is musubi tuner speed with 7 block swap. speeds are almost same

<img width="3445" height="930" alt="Image" src="https://github.com/user-attachments/assets/9572b2d8-fdd7-404c-94b8-9b908fb7dc6e" />

**Comment by xzuyn** [comment]
Should post exact configs to be able to compare.

**Comment by kohya-ss** [comment]
When training LoRA with rank (dim) 4 with Musubi Tuner, VRAM usage is around 25GB. This means that Musubi Tuner uses more VRAM when increasing the rank.

The exact reasons are unknown, so this is just a guess, but it may be due to different handling of bf16. Even when using mixed precision training with Musubi Tuner, LoRA weights are stored in float32 precision (unlike sd-scripts, Musubi Tuner does not support full_bf16 for LoRA training). In OneTrainer, they may be stored in bf16 by default, and stochastic rounding may be combined.

This is intentional to maintain accuracy, so there are currently no plans to support full_bf16 for LoRA training with Musubi Tuner. I think rank 128 is probably too large for general usage.

**Comment by sdbds** [comment]
> When training LoRA with rank (dim) 4 with Musubi Tuner, VRAM usage is around 25GB. This means that Musubi Tuner uses more VRAM when increasing the rank.
> 
> The exact reasons are unknown, so this is just a guess, but it may be due to different handling of bf16. Even when using mixed precision training with Musubi Tuner, LoRA weights are stored in float32 precision (unlike sd-scripts, Musubi Tuner does not support full_bf16 for LoRA training). In OneTrainer, they may be stored in bf16 by default, and stochastic rounding may be combined.
> 
> This is intentional to maintain accuracy, so there are currently no plans to support full_bf16 for LoRA training with Musubi Tuner. I think rank 128 is probably too large for general usage.

Since the SD era, even using full bf16 for LoRa training has little impact on the results (even without random rounding), only having a significant impact on fine-tuning.
If it's just LoRa, I think we could try supporting full BF16.

**Comment by FurkanGozukara** [comment]
> When training LoRA with rank (dim) 4 with Musubi Tuner, VRAM usage is around 25GB. This means that Musubi Tuner uses more VRAM when increasing the rank.
> 
> The exact reasons are unknown, so this is just a guess, but it may be due to different handling of bf16. Even when using mixed precision training with Musubi Tuner, LoRA weights are stored in float32 precision (unlike sd-scripts, Musubi Tuner does not support full_bf16 for LoRA training). In OneTrainer, they may be stored in bf16 by default, and stochastic rounding may be combined.
> 
> This is intentional to maintain accuracy, so there are currently no plans to support full_bf16 for LoRA training with Musubi Tuner. I think rank 128 is probably too large for general usage.

this makes total sense. this is probably the reason 

would it be too hard to implement? perhaps we can reduce vram and speed up with no quality loss

yes they have stochastic rounding

**Comment by dxqb** [comment]
> When training LoRA with rank (dim) 4 with Musubi Tuner, VRAM usage is around 25GB. This means that Musubi Tuner uses more VRAM when increasing the rank.
> 
> The exact reasons are unknown, so this is just a guess, but it may be due to different handling of bf16. Even when using mixed precision training with Musubi Tuner, LoRA weights are stored in float32 precision (unlike sd-scripts, Musubi Tuner does not support full_bf16 for LoRA training). In OneTrainer, they may be stored in bf16 by default, and stochastic rounding may be combined.
> 

default in OneTrainer is also float32 LoRA weights
bf16 is possible but I don't know that I'd recommend that

**Comment by dxqb** [comment]
I can also only speculate but is beta1 == 0 when you use Adafactor by default or do you have to set that?
beta1 > 0 with adafactor takes another parameter buffer full of vram, so it also increases with rank
